src/main/java/ch/njol/skript/expressions/ExprSets.java

```diff
@@ -39,152 +29,80 @@
+			return false;
 		}
 		return true;
 	}
 
-	private Object[] buffer = null;
-
 	@Override
 	protected Object[] get(Event event) {
-		if (buffer != null)
-			return buffer;
-		List<Object> elements = new ArrayList<>();
-		for (Object element : new IteratorIterable<>(iterator(event)))
-			elements.add(element);
-		if (types instanceof Literal)
-			return buffer = elements.toArray();
-		return elements.toArray();
+		assert supplier != null;
+		Iterator<?> iterator = supplier.get();
+		List<?> elements = Lists.newArrayList(iterator);
+		return elements.toArray(new Object[0]);
 	}
 
 	@Override
 	@Nullable
-	public Iterator<Object> iterator(Event event) {
-		if (pattern == 4)
-			return new ArrayIterator<>(SkriptColor.values());
-		if (pattern == 0 || pattern == 3)
-			return new Iterator<Object>() {
-				
-				private final Iterator<Material> iter = new ArrayIterator<>(Material.values());
-				
-				@Override
-				public boolean hasNext() {
-					return iter.hasNext();
-				}
-				
-				@Override
-				public ItemStack next() {
-					return new ItemStack(iter.next());
-				}
-				
-				@Override
-				public void remove() {}
-				
-			};
-		Iterator<ItemType> it = new ArrayIterator<>(types.getArray(event));
-		if (!it.hasNext())
-			return null;
-		Iterator<Object> iter;
-		iter = new Iterator<Object>() {
-			
-			Iterator<ItemStack> current = it.next().getAll().iterator();
-			
-			@Override
-			public boolean hasNext() {
-				while (!current.hasNext() && it.hasNext()) {
-					current = it.next().getAll().iterator();
-				}
-				return current.hasNext();
-			}
-			
-			@Override
-			public ItemStack next() {
-				if (!hasNext())
-					throw new NoSuchElementException();
-				return current.next();
-			}
-			
-			@Override
-			public void remove() {}
-			
-		};
-
-		return new CheckedIterator<Object>(iter, new NullableChecker<Object>() {
-			@Override
-			public boolean check(@Nullable Object ob) {
-				if (ob == null)
-					return false;
-				if (ob instanceof ItemStack)
-					if (!((ItemStack) ob).getType().isBlock())
-						return false;
-				return true;
-			}
-		});
-	}
-	
-	@Override
-	public Class<? extends Object> getReturnType() {
-		if (pattern == 4)
-			return Color.class;
-		return ItemStack.class;
-	}
-	
-	@Override
-	public String toString(@Nullable Event event, boolean debug) {
-		if (event == null)
-			return "sets of data. Pattern " + pattern;
-		if (pattern == 4)
-			return "colours";
-		return (pattern < 2 ? "blocks" : "items") + (types != null ? " of type" + (types.isSingle() ? "" : "s") + " " + types.toString(event, debug) : "");
+	public Iterator<?> iterator(Event event) {
+		assert supplier != null;
+		return supplier.get();
 	}
-	
+
 	@Override
 	public boolean isSingle() {
 		return false;
 	}
-	
+
 	@Override
-	public boolean isLoopOf(String s) {
-		return pattern == 4 && (s.equalsIgnoreCase("color") || s.equalsIgnoreCase("colour"))|| pattern >= 2 && s.equalsIgnoreCase("block") || pattern < 2 && s.equalsIgnoreCase("item");
```

Comment:
I think we could support isLoopOf here by parsing the input as a classinfo and then seeing if the resulting classinfo (if any) is assignable from the classinfo literal's class